types/coin.go

```diff
@@ -172,6 +172,29 @@ func (coins Coins) IsValid() bool {
 	}
 }
 
+// AddCoinByDenom looks for a matching coin in coins that has the same denom as
+// other. Upon matching, the other coin is added to the matching coin. If the
+// coins are empty, the coin is simply added to the empty set. If the denom of
+// the coin to be added does not exist in coins, the function will panic.
+func (coins Coins) AddCoinByDenom(other Coin) Coins {
+	res := copyCoins(coins)
+
+	if len(res) == 0 {
+		return Coins{other}
+	}
+
+	// TODO: Perform binary search on coins to improve runtime performance
+	// (i.e. implement a generic binary coin search on Coins)
+	for i, coin := range res {
```

Comment:
AddCoinByDenom shouldn't panic when coin doesn't exist in coins.  The name of the function and the scope of functionality implies that it shouldn't.  Also it's not even true, when len(res) == 0, the comments are wrong.  This function shouldn't panic.

Reply:
Hmmm you actually suggested the `SubCoinByDenom` variant SHOULD panic here: https://github.com/cosmos/cosmos-sdk/pull/2694#discussion_r239417664. Why not the `AddCoinByDenom` variant?

Reply:
Because there's no reason for this function to panic when implemented correctly.

Reply:
Sure! So do you suggest we simply return `coins` when the denom isn't present? What's wrong with the implementation?

Reply:
Updated :)